### ISSUE TITLE
[minecraft] Evaluate JVM G1 Garbage Collector

### ISSUE THREAD

**Issue by zaquestion** [itzg/dockerfiles#126]
Would you mind switching or maintaining a second `dockerfile` using the jdk image on https://hub.docker.com/r/library/openjdk/tags/

Its _only_ 100MB more space, my vote would be to just cut over, but a seperate tag/dockerfile would be excellent

ref on jvm options that require jdk
http://gaming.stackexchange.com/questions/104527/how-can-i-run-a-minecraft-server-on-multiple-cpu-cores

Only requires changing `jre` to `jdk` here: https://github.com/itzg/dockerfiles/blob/master/minecraft-server/Dockerfile#L1

FWIW, I built my own version of this locally and was able to start up both a vanilla and forge based server

**Comment by itzg** [comment]
I agree that's a reasonable increase in size; however, having worked with Java for many years now I am quite certain the JDK doesn't add any JVM options beyond the JRE. The [official faq](https://www.java.com/en/download/faq/techinfo.xml) confirms my understanding that the JDK differs by including development-time tools such as `javac`, `jar`, etc.

I did notice that I haven't enabled the [G1 garbage collector](https://www.java.com/en/download/faq/techinfo.xml) which would provide much more improvements for large memory, multi-core hosting of Minecraft.

I can go ahead and make the G1 GC change, but let me know if you can find any official docs on multi-core differences of the JDK vs JRE.

**Comment by itzg** [comment]
@zaquestion , can you give the image I just pushed, `50e62877c056`, a try? I'm also going to stress test it tomorrow on my dual CPU, quad core system.

**Comment by itzg** [comment]
...I also introduced a new `-e MAX_MEMORY=...` option to make it easier to bump up the JVM memory limit.

**Comment by zaquestion** [comment]
@itzg I'm trying this out, so far so good, I also used your `MAX_MEMORY` option. I am seeing a lot more JVM memory getting used and a lot less garbage collects which seems nice.

Here is the config
```
  infinity:
    image: itzg/minecraft-server
    restart: always
    volumes:
      - /minecraft_infinity:/data
      - /storage/minecraft_infinity/plugins/dynmap:/data/plugins/dynmap
    ports:
      - 25565:25565
      - 80:8123
      - 9225:9225
    environment:
      - TYPE=SPIGOT
      - CONSOLE=false
      - EULA=TRUE
      - LEVEL_TYPE=LARGEBIOMES
      - PVP=TRUE
      - VIEW_DISTANCE=16
      - MAX_BUILD_HEIGHT=512
      - MAX_PLAYERS=100
      - OPS=zaquestion
      - DIFFICULTY=hard
      - MAX_MEMORY=8G
```

**Comment by itzg** [comment]
@zaquestion thanks for trying it out and the progress update. I haven't done my large scale testing yet, so I'm especially glad you are progressing.

**Comment by itzg** [comment]
@zaquestion with some experimentation (and remembering an issue others brought up a while ago) I found that this worked well with a 32GB server and render distance of 15:

https://github.com/itzg/dockerfiles/blob/master/minecraft-server/docker-compose-big.yml

The `CONSOLE: "false"` [here](https://github.com/itzg/dockerfiles/blob/master/minecraft-server/docker-compose-big.yml#L16) seemed to be the critical part to avoid maxing out a core

**Comment by zaquestion** [comment]
Following up here. I've been running with these changes for almost a couple weeks. Its hard to quantify even with various charts in grafana. A general look at CPU usage before and after enabling G1 shows a lot CPU usage on avg. I haven't seen any single core maxed out. If I had to **guess** the GC is cleaning up everything so quickly in parallel its not quite getting recognized prometheus/grafana. 

On bumping the max memory.
I tried setting `MAX_MEMORY` to 8G, 32G, and 64G at each level the results were similar. On my server about 1G of memory is actually used. Which ( and I'm not a Java guy) means when the initial memory pool is not set (Xms) the pool will only grow about that large (or less) yielding very frequent GCs. When Xms is set to the `MAX_MEMORY` value, garbage will collect (rather quickly in proportion to available memory) until about `MAX_MEMORY / 2` and the GC will run. However I

I use https://github.com/sladkoff/minecraft-prometheus-exporter to get various metrics out of minecraft and gives me a good look at when the GC is running over time.

And https://www.spigotmc.org/resources/react-smart-server-performance.21057/ to monitor similar metrics in game, but it also exposes how much memory minecraft is actually using and how much is straight up garbage. 

TL;DR
More memory means less GC, but more wasted memory.
The G1 GC seems so quickly resolve 10s of Gigs of garbage without a noticeable blip ingame or on the charts.

**Comment by itzg** [comment]
Thanks for the great write-up. I'm glad to hear you're observing the same positive benefits of G1 that I have seen in a similar switch we did at work for our production Cassandra cluster.

As for `MAX_MEMORY`, the startup script does also use that value for `-Xms` [here](https://github.com/itzg/dockerfiles/blob/master/minecraft-server/start-minecraft.sh#L496). The discrepancy seems to be the usual mystery difference/vagueness between the various meanings of "allocated memory" :)

Thanks for the pointers about the tools. I recently bumped into prometheus and that's cool to see someone added a Minecraft integration.

Yes, your summary is also what I have seen and heard in general about Java heap allocation and GC trade-off.

So, all in all it sounds like a success. Thanks for your help on this one.

**Comment by aamche** [comment]
In simple terms, why would it be good to set Xms to the same as Xmx. I've generally set Xms to half of Xmx, as it allows me to run more instances. Seems we've lost memory control by adding MEMORY into the environment args.

**Comment by itzg** [comment]
@aamche excellent point. The motivation was to avoid performance degradation while a server "warmed up"; however, in professional Java career I rarely set Xms since I rarely see heap expansion as a performance issue.

You have inspired me to change `MEMORY` to only affect Xmx and I'll introduce a companion variable to tweak the initial heap size.

**Comment by aamche** [comment]
Thanks itzg, I know you provide all this for free, so thank you for all your hard work. 

Memory is a funny one,  in my setup I have bungeecord linking to various games, so it's often a few of them are empty at any one time.

**Comment by itzg** [comment]
@zaquestion and @aamche , can you take a look at the [updated README section](https://github.com/itzg/dockerfiles/tree/master/minecraft-server#memory-limit) and see if that makes sense for the various scenarios? I ultimately decided to leave `MEMORY` as the one-stop for setting both initial and max since that approach has the "least surprises". But @aamche for scenarios like you describe, you can tweak `INIT_MEMORY` to be lower.

I have pushed to Hub, so `docker pull` to try out these changes.

**Comment by aamche** [comment]
Looks great, thanks itzg. Things are working smoothly and ram is all in good again. The swap was getting hammered for a while. Easy implementation too.

**Comment by itzg** [comment]
Thanks @aamche
